src/test/java/teammates/test/cases/testdriver/TestngTest.java

```diff
@@ -0,0 +1,115 @@
+        return FileHelper.readFile(file);
+    }
+    
+    /**
+     * Files to be checked in testng.xml are added to testFiles
+     * 
+     * @param testNgXml    Contents of testng.xml
+     * @return             HashMap containing <class name, package name>
+     */
+    private HashMap<String, String> getTestFiles(String testNgXml) {       
+        // BaseComponentTestCase, BaseTestCase (files in current directory) excluded because 
+        // base classes are extended by the actual tests
+        
+        return addFilesToTestsRecursively(new HashMap<String, String>(), 
+                                              "./src/test/java/teammates/test/cases", true, "", testNgXml); 
+    }
+    
+    /**
+     * Exclude files which do not have tests
+     * 
+     * @param testFiles    Files to be checked before excluding tests
+     * @return             Files to be checked after excluding tests
+     */
+    private HashMap<String, String> excludeFiles(HashMap<String, String> testFiles) {
+        // Base*TestCase are base classes to be extended by the actual tests
+        testFiles.remove("BaseUiTestCase");
+        
+        // FeedbackQuestionUiTest is the base class for all Feedback*QuestionUiTest (different question types)
+        testFiles.remove("FeedbackQuestionUiTest");
+        
+        // Needs to be run only when changes are made to GodMode
+        testFiles.remove("GodModeTest");
+        
+        return testFiles;
+    }
+    
+    private boolean isTestFileIncluded(String testNgXml, Entry<String, String> testFileName) {
+        return testNgXml.contains("<class name=\"teammates.test.cases" 
+                                       + testFileName.getValue() + "." + testFileName.getKey() + "\" />");
+    }
+
+    /**
+     * Recursively add files from testng.xml which are to be checked
+     * 
+     * @param testFiles                       HashMap containing <class name, package name>
+     * @param path                            Check files and directories in the current path
+     * 
+     * @param areFilesInCurrentDirExcluded    If true, files in the current path are not
+     *                                        added to tests but sub-directories are checked
+     *                                        
+     * @param packageName                     Package name of the current file
+     * @param testNgXml                       Contents of testng.xml
+     * 
+     * @return                                HashMap containing <class name, package name> including 
+     *                                        current file or tests in the current directory
+     */
+    private HashMap<String, String> addFilesToTestsRecursively(HashMap<String, String> testFiles, String path, 
+                                                               boolean areFilesInCurrentDirExcluded, 
+                                                               String packageName, String testNgXml) {
```

Comment:
This method is checking if the path refers to a file or directory. If it is a file, it is added to tests. If it is a directory, it calls the method recursively to check that directory. It does not list all files in the directory, only adds it to the tests. Also, there are some directories where the files are not checked but the sub-directories are checked, so it is hard to separate.